storage/BloomFilterIndexSubBlock.cpp

```diff
@@ -0,0 +1,271 @@
+  // Note: This function is only invoked when EstimateBytesPerTuple() returns zero.
+  DEBUG_ASSERT(DescriptionIsValid(relation, description));
+  return description.GetExtension(BloomFilterIndexSubBlockDescription::bloom_filter_size);
+}
+
+bool BloomFilterIndexSubBlock::addEntry(const tuple_id tuple) {
+  DCHECK(is_initialized_);
+  if (!is_consistent_) {
+    return false;
+  }
+  if (!tuple_store_.hasTupleWithID(tuple)) {
+    return false;
+  }
+  for (std::size_t i = 0; i < indexed_attribute_ids_.size(); ++i) {
+    const attribute_id indexed_attribute_id = indexed_attribute_ids_[i];
+    const TypedValue tuple_value = tuple_store_.getAttributeValueTyped(tuple, indexed_attribute_id);
+    const std::uint8_t *data = static_cast<const std::uint8_t*>(tuple_value.getDataPtr());
+    const std::size_t data_size = tuple_value.getDataSize();
+    if (!tuple_value.isNull()) {
+      bloom_filter_->insert(data, data_size);
+    }
+  }
+  return true;
+}
+
+bool BloomFilterIndexSubBlock::bulkAddEntries(const TupleIdSequence &tuples) {
+  DCHECK(is_initialized_);
+  if (!is_consistent_) {
+    return false;
+  }
+  bool didAllSucceed = true;
+  for (const tuple_id &tuple : tuples) {
+    didAllSucceed = didAllSucceed && addEntry(tuple);
+  }
+  return didAllSucceed;
+}
+
+void BloomFilterIndexSubBlock::removeEntry(const tuple_id tuple) {
+  is_consistent_ = false;
+}
+
+void BloomFilterIndexSubBlock::bulkRemoveEntries(const TupleIdSequence &tuples) {
+  is_consistent_ = false;
+}
+
+predicate_cost_t BloomFilterIndexSubBlock::estimatePredicateEvaluationCost(
+    const ComparisonPredicate &predicate) const {
+  DCHECK(is_initialized_);
+  BloomFilterSelectivity selectivity = getSelectivityForPredicate(predicate);
+  // Note: A Bloomfilter index is only useful when it gives a zero selectivity
+  //       in which case a block can be skipped entirely.
+  if (selectivity == BloomFilterSelectivity::kSelectivityNone) {
+    // Return minimum cost so that this block can be skipped altogether.
+    return predicate_cost::kConstantTime;
+  } else {
+    // Return maximum cost because this is as worse as doing a full scan of the block.
+    return predicate_cost::kInfinite;
+  }
+}
+
+TupleIdSequence* BloomFilterIndexSubBlock::getMatchesForPredicate(
+    const ComparisonPredicate &predicate,
+    const TupleIdSequence *filter) const {
+  DCHECK(is_initialized_);
+  if (filter != nullptr) {
+    LOG(FATAL) << "BloomFilterIndex does not support filter evaluation with predicate.";
+  }
+  BloomFilterSelectivity selectivity = getSelectivityForPredicate(predicate);
+  if (selectivity == BloomFilterSelectivity::kSelectivityNone) {
+    // Best-case scenario, when a BloomFilter index allows to skip a block entirely.
+    // A new tuple ID sequence is initialized to false for all values.
+    return new TupleIdSequence(tuple_store_.numTuples());
+  } else if (selectivity == BloomFilterSelectivity::kSelectivityAll) {
```

Comment:
Oh I see, the definite for `all` isn't exactly what I thought it meant. It seems to mean that:

```c++
 // Bloom filter hit. Can be a false positive, a scan is required here.
 return BloomFilterSelectivity::kSelectivityAll;
```